README.md
<!-- PROJECT LOGO -->
<p align="right">
<a href="https://pub.dev">
<img src="https://raw.githubusercontent.com/born-ideas/measurements/master/assets/project_badge.png" height="100" alt="badge">
</a>
</p>
<p align="center">
<img src="https://raw.githubusercontent.com/born-ideas/measurements/master/assets/project_logo.png" height="100" alt="logo" />
</p>

<!-- PROJECT SHIELDS -->
<p align="center">
<a href="https://pub.dev/packages/measurements"><img src="https://img.shields.io/pub/v/measurements" alt="pub"></a>
<a href="https://github.com/born-ideas/measurements/actions?query=workflow%3Atest"><img src="https://img.shields.io/github/workflow/status/born-ideas/measurements/test?label=test" alt="test"></a>
<a href="https://github.com/born-ideas/measurements/issues"><img src="https://img.shields.io/github/issues/born-ideas/measurements" alt="issues"></a>
<a href="https://github.com/born-ideas/measurements/network"><img src="https://img.shields.io/github/forks/born-ideas/measurements" alt="forks"></a>
<a href="https://github.com/born-ideas/measurements/stargazers"><img src="https://img.shields.io/github/stars/born-ideas/measurements" alt="stars"></a>
<a href="https://dart.dev/guides/language/effective-dart/style"><img src="https://img.shields.io/badge/style-effective_dart-40c4ff.svg" alt="style"></a>
<a href="https://github.com/born-ideas/measurements/blob/master/LICENSE"><img src="https://img.shields.io/github/license/born-ideas/measurements" alt="license"></a>
</p>

---

<!-- TABLE OF CONTENTS -->
## Table of Contents
* [About the Project](#about-the-project)
* [Getting Started](#getting-started)
* [Usage](#usage)
* [Roadmap](#roadmap)
* [Contributing](#contributing)
* [License](#license)
* [Contact](#contact)
* [Acknowledgements](#acknowledgements)



<!-- ABOUT THE PROJECT -->
## About The Project
<p align="center">
<img src="https://raw.githubusercontent.com/born-ideas/measurements/master/assets/screenshot_1.gif" width="600" alt="Screenshot 1" />
</p>

Working with different units of measurements can be a pain. This project provides a [Dart Library](https://pub.dev/packages/measurements) 
with basic classes for working with various measurement units. So far the supported measurements include:
- [x] Length
- [x] Mass
- [x] Area

### Built With
* [Dart](https://dart.dev/)



<!-- GETTING STARTED -->
## Getting Started
### Prerequisites
This package can only be used by other dart projects. If this is your first dart project, see the following pages to help you get started:                   
- [Codelabs: Guided, hands-on coding experience](https://dart.dev/codelabs)
- [Tutorials: Teaching you how to use Dart](https://dart.dev/tutorials)

### Installation
Add `measurements` as a [dependency in your pubspec.yaml file](https://dart.dev/guides/packages).



<!-- USAGE EXAMPLES -->
## Usage
Conversions:
```dart
import 'package:measurements/measurements.dart';

void main(List<String> arguments) async {
  Length heightOfEverest = Length.fromMetres(8848.0);
  print("Height of Everest (in metres): ${heightOfEverest.inMetres}.");
  print("Height of Everest (in kilometres): ${heightOfEverest.inKilometres}.");
  print("Height of Everest (in yards): ${heightOfEverest.inYards}.");
  print("Height of Everest (in miles): ${heightOfEverest.inMiles}.");
}
```

Calculations:
```dart
import 'package:measurements/measurements.dart';

void main(List<String> arguments) async {
  Area a = Area.fromAcres(87.0);
  Area b = Area.fromAcres(43.0);
  print("a + b is ${(a + b).inAcres} acres.");
  print("a - b is ${(a - b).inAcres} acres.");
  print("a * 4 is ${(a * 4).inAcres} acres.");
  print("a / 4 is ${(a / 4).inAcres} acres.");
}
```

Comparisons:
```dart
import 'package:measurements/measurements.dart';

void main(List<String> arguments) async {
  Mass elephantWeight = Mass.fromKilograms(5400);
  Mass zebraWeight = Mass.fromKilograms(380);
  if (elephantWeight == zebraWeight) {
    print("An elephant and a Zebra have the same weight.");
  } else if (zebraWeight < elephantWeight) {
    Mass difference = elephantWeight - zebraWeight;
    print("An elephant is ${difference.inPounds} pounds heavier than a zebra.");
  } else {
    Mass difference = zebraWeight - elephantWeight;
    print("A Zebra is ${difference.inPounds} pounds heavier than an elephant.");
  }
}
```



<!-- ROADMAP -->
## Roadmap
We are hoping to provide the following measurements in the near future:
- [ ] DataTransferRate
- [ ] DigitalStorage
- [ ] Energy
- [ ] Frequency
- [ ] Pressure
- [ ] Speed
- [ ] Temperature
- [ ] Volume

See the [open issues](https://github.com/othneildrew/Best-README-Template/issues) for a list of other proposed features (and known issues).



<!-- CONTRIBUTING -->
## Contributing

Contributions are what make the open source community such an amazing place to be learn, inspire, and create. Any contributions you make are **greatly appreciated**.

1. Fork the Project
2. Create your Feature Branch (`git checkout -b feature/AmazingFeature`)
3. Commit your Changes (`git commit -m 'Add some AmazingFeature'`)
4. Push to the Branch (`git push origin feature/AmazingFeature`)
5. Open a Pull Request



<!-- LICENSE -->
## License

Distributed under the MIT License. See [LICENSE](LICENSE) for more information.



<!-- CONTACT -->
## Contact

BornIdeas - [born.dev](https://www.born.dev) - [info@born.dev](mailto:support@born.dev)

Project Link: [https://github.com/born-ideas/measurements](https://github.com/born-ideas/measurements)



<!-- ACKNOWLEDGEMENTS -->
## Acknowledgements
* [Shields IO](https://shields.io)
* [Open Source Licenses](https://choosealicense.com)

CHANGELOG.md
## 0.0.6

* Added zero-valued constants for each measurement

## 0.0.5

* Fixed example project

## 0.0.4

* Documentation updates

## 0.0.3

* Added support for `Area` measurements

## 0.0.2

* Corrected ounces for `Mass` class
* Fixed overloaded `==` operator

## 0.0.1

* Initial Release

test/area_test.dart
import 'package:measurements/measurements.dart';
import 'package:test/test.dart';

void main() {
  group("Area", () {
    test("fromSquareCentimetres", () {
      Area area = Area.fromSquareCentimetres(10000000000);

      expect(area.inSquareMetres, 1000000);
      expect(area.inSquareKilometres, 1);
      expect(area.inHectares, 100);
      expect(area.inSquareInches, 25000000000000 / 16129);
      expect(area.inSquareFeet, 1562500000000 / 145161);
      expect(area.inSquareYards, 1562500000000 / 1306449);
      expect(area.inSquareMiles, 244140625 / 632321316);
      expect(area.inAcres, 3088817268339567 / 12500000000000);
    });

    test("fromSquareMetres", () {
      Area area = Area.fromSquareMetres(1000000);

      expect(area.inSquareCentimetres, 10000000000);
      expect(area.inSquareKilometres, 1);
      expect(area.inHectares, 100);
      expect(area.inSquareInches, 25000000000000 / 16129);
      expect(area.inSquareFeet, 1562500000000 / 145161);
      expect(area.inSquareYards, 1562500000000 / 1306449);
      expect(area.inSquareMiles, 244140625 / 632321316);
      expect(area.inAcres, 3088817268339567 / 12500000000000);
    });

    test("fromSquareKilometres", () {
      Area area = Area.fromSquareKilometres(1);

      expect(area.inSquareCentimetres, 10000000000);
      expect(area.inSquareMetres, 1000000);
      expect(area.inHectares, 100);
      expect(area.inSquareInches, 25000000000000 / 16129);
      expect(area.inSquareFeet, 1562500000000 / 145161);
      expect(area.inSquareYards, 1562500000000 / 1306449);
      expect(area.inSquareMiles, 244140625 / 632321316);
      expect(area.inAcres, 3088817268339567 / 12500000000000);
    });

    test("fromHectares", () {
      Area area = Area.fromHectares(100);

      expect(area.inSquareCentimetres, 10000000000);
      expect(area.inSquareMetres, 1000000);
      expect(area.inSquareKilometres, 1);
      expect(area.inSquareInches, 25000000000000 / 16129);
      expect(area.inSquareFeet, 1562500000000 / 145161);
      expect(area.inSquareYards, 1562500000000 / 1306449);
      expect(area.inSquareMiles, 244140625 / 632321316);
      expect(area.inAcres, 3088817268339567 / 12500000000000);
    });

    test("fromSquareInches", () {
      Area area = Area.fromSquareInches(25000000000000 / 16129);

      expect(area.inSquareCentimetres, 10000000000);
      expect(area.inSquareMetres, 1000000);
      expect(area.inSquareKilometres, 1);
      expect(area.inHectares, 100);
      expect(area.inSquareFeet, 1562500000000 / 145161);
      expect(area.inSquareYards, 1562500000000 / 1306449);
      expect(area.inSquareMiles, 244140625 / 632321316);
      expect(area.inAcres, 3088817268339567 / 12500000000000);
    });

    test("fromSquareFeet", () {
      Area area = Area.fromSquareFeet(1562500000000 / 145161);

      expect(area.inSquareCentimetres, 10000000000);
      expect(area.inSquareMetres, 1000000);
      expect(area.inSquareKilometres, 1);
      expect(area.inHectares, 100);
      expect(area.inSquareInches, 25000000000000 / 16129);
      expect(area.inSquareYards, 1562500000000 / 1306449);
      expect(area.inSquareMiles, 244140625 / 632321316);
      expect(area.inAcres, 3088817268339567 / 12500000000000);
    });

    test("fromSquareYards", () {
      Area area = Area.fromSquareYards(1562500000000 / 1306449);

      expect(area.inSquareCentimetres, 10000000000);
      expect(area.inSquareMetres, 1000000);
      expect(area.inSquareKilometres, 1);
      expect(area.inHectares, 100);
      expect(area.inSquareInches, 25000000000000 / 16129);
      expect(area.inSquareFeet, 1562500000000 / 145161);
      expect(area.inSquareMiles, 244140625 / 632321316);
      expect(area.inAcres, 3088817268339567 / 12500000000000);
    });

    test("fromSquareMiles", () {
      Area area = Area.fromSquareMiles(244140625 / 632321316);

      expect(area.inSquareCentimetres, 10000000000);
      expect(area.inSquareMetres, 1000000);
      expect(area.inSquareKilometres, 1);
      expect(area.inHectares, 100);
      expect(area.inSquareInches, 25000000000000 / 16129);
      expect(area.inSquareFeet, 1562500000000 / 145161);
      expect(area.inSquareYards, 1562500000000 / 1306449);
      expect(area.inAcres, 3088817268339567 / 12500000000000);
    });

    test("fromAcres", () {
      Area area = Area.fromAcres(3088817268339567 / 12500000000000);

      expect(area.inSquareCentimetres, 10000000000);
      expect(area.inSquareMetres, 1000000);
      expect(area.inSquareKilometres, 1);
      expect(area.inHectares, 100);
      expect(area.inSquareInches, 25000000000000 / 16129);
      expect(area.inSquareFeet, 1562500000000 / 145161);
      expect(area.inSquareYards, 1562500000000 / 1306449);
      expect(area.inSquareMiles, 244140625 / 632321316);
    });

    test("fromRectangle", () {
      Length length = Length.fromCentimetres(5);
      Length width = Length.fromCentimetres(8);
      Area area = Area.fromRectangle(length, width);

      expect(area.inSquareCentimetres, 40);
    });

    test("fromTriangle", () {
      Length base = Length.fromMetres(6);
      Length height = Length.fromMetres(4);
      Area area = Area.fromTriangle(base, height);

      expect(area.inSquareMetres, 12);
    });

    test("fromCircle", () {
      Length radius = Length.fromKilometres(4);
      Area area = Area.fromCircle(radius);

      expect(area.inSquareKilometres, 50.26548245743669);
    });
  });
}


test/mass_test.dart
import 'package:measurements/measurements.dart';
import 'package:test/test.dart';

void main() {
  group("Mass", (){
    test("fromMicrograms", (){
      Mass testMass = Mass.fromMicrograms(1000000000);

      expect(testMass.inNanograms, 1000000000000);
      expect(testMass.inMilligrams, 1000000);
      expect(testMass.inGrams, 1000);
      expect(testMass.inKilograms, 1);
      expect(testMass.inMetricTonnes, 0.001);
      expect(testMass.inOunces, 35.27396194958041);
      expect(testMass.inStones, 0.15747304441776971);
      expect(testMass.inPounds, 2.2046226218487757);
      expect(testMass.inImperialTonnes, 0.0009842065276110606);
      expect(testMass.inUSTonnes, 0.001102311310924388);
    });

    test("fromMilligrams", (){
      Mass testMass = Mass.fromMilligrams(1000000);

      expect(testMass.inNanograms, 1000000000000);
      expect(testMass.inMicrograms, 1000000000);
      expect(testMass.inGrams, 1000);
      expect(testMass.inKilograms, 1);
      expect(testMass.inMetricTonnes, 0.001);
      expect(testMass.inOunces, 35.27396194958041);
      expect(testMass.inStones, 0.15747304441776971);
      expect(testMass.inPounds, 2.2046226218487757);
      expect(testMass.inImperialTonnes, 0.0009842065276110606);
      expect(testMass.inUSTonnes, 0.001102311310924388);
    });

    test("fromGrams", (){
      Mass testMass = Mass.fromGrams(1000);

      expect(testMass.inNanograms, 1000000000000);
      expect(testMass.inMicrograms, 1000000000);
      expect(testMass.inMilligrams, 1000000);
      expect(testMass.inKilograms, 1);
      expect(testMass.inMetricTonnes, 0.001);
      expect(testMass.inOunces, 35.27396194958041);
      expect(testMass.inStones, 0.15747304441776971);
      expect(testMass.inPounds, 2.2046226218487757);
      expect(testMass.inImperialTonnes, 0.0009842065276110606);
      expect(testMass.inUSTonnes, 0.001102311310924388);
    });

    test("fromKilograms", (){
      Mass testMass = Mass.fromKilograms(1);

      expect(testMass.inNanograms, 1000000000000);
      expect(testMass.inMicrograms, 1000000000);
      expect(testMass.inMilligrams, 1000000);
      expect(testMass.inGrams, 1000);
      expect(testMass.inMetricTonnes, 0.001);
      expect(testMass.inOunces, 35.27396194958041);
      expect(testMass.inStones, 0.15747304441776971);
      expect(testMass.inPounds, 2.2046226218487757);
      expect(testMass.inImperialTonnes, 0.0009842065276110606);
      expect(testMass.inUSTonnes, 0.001102311310924388);
    });

    test("fromTonnes", (){
      Mass testMass = Mass.fromMetricTonnes(0.001);

      expect(testMass.inNanograms, 1000000000000);
      expect(testMass.inMicrograms, 1000000000);
      expect(testMass.inMilligrams, 1000000);
      expect(testMass.inGrams, 1000);
      expect(testMass.inKilograms, 1);
      expect(testMass.inOunces, 35.27396194958041);
      expect(testMass.inStones, 0.15747304441776971);
      expect(testMass.inPounds, 2.2046226218487757);
      expect(testMass.inImperialTonnes, 0.0009842065276110606);
      expect(testMass.inUSTonnes, 0.001102311310924388);
    });

    test("fromOunces", (){
      Mass testMass = Mass.fromOunces(16);

      expect(testMass.inNanograms, 453592370000);
      expect(testMass.inMicrograms, 453592370);
      expect(testMass.inMilligrams, 453592.37);
      expect(testMass.inGrams, 453.59237);
      expect(testMass.inKilograms, 0.45359237);
      expect(testMass.inMetricTonnes, 0.00045359237);
      expect(testMass.inStones, 0.07142857142857142);
      expect(testMass.inPounds, 1);
      expect(testMass.inImperialTonnes, 0.0004464285714285714);
      expect(testMass.inUSTonnes, 0.0005);
    });

    test("fromStones", (){
      Mass testMass = Mass.fromStones(0.07142857142857142);

      expect(testMass.inNanograms, 453592370000);
      expect(testMass.inMicrograms, 453592370);
      expect(testMass.inMilligrams, 453592.37);
      expect(testMass.inGrams, 453.59237);
      expect(testMass.inKilograms, 0.45359237);
      expect(testMass.inMetricTonnes, 0.00045359237);
      expect(testMass.inOunces, 16);
      expect(testMass.inPounds, 1);
      expect(testMass.inImperialTonnes, 0.0004464285714285714);
      expect(testMass.inUSTonnes, 0.0005);
    });

    test("fromPounds", (){
      Mass testMass = Mass.fromPounds(1);

      expect(testMass.inNanograms, 453592370000);
      expect(testMass.inMicrograms, 453592370);
      expect(testMass.inMilligrams, 453592.37);
      expect(testMass.inGrams, 453.59237);
      expect(testMass.inKilograms, 0.45359237);
      expect(testMass.inMetricTonnes, 0.00045359237);
      expect(testMass.inOunces, 16);
      expect(testMass.inStones, 0.07142857142857142);
      expect(testMass.inImperialTonnes, 0.0004464285714285714);
      expect(testMass.inUSTonnes, 0.0005);
    });

    test("fromImperialTonnes", (){
      Mass testMass = Mass.fromImperialTonnes(0.0004464285714285714);

      expect(testMass.inNanograms, 453592370000);
      expect(testMass.inMicrograms, 453592370);
      expect(testMass.inMilligrams, 453592.37);
      expect(testMass.inGrams, 453.59237);
      expect(testMass.inKilograms, 0.45359237);
      expect(testMass.inMetricTonnes, 0.00045359237);
      expect(testMass.inOunces, 16);
      expect(testMass.inStones, 0.07142857142857142);
      expect(testMass.inPounds, 1);
      expect(testMass.inUSTonnes, 0.0005);
    });

    test("fromUSTonnes", (){
      Mass testMass = Mass.fromUSTonnes(0.0005);

      expect(testMass.inNanograms, 453592370000);
      expect(testMass.inMicrograms, 453592370);
      expect(testMass.inMilligrams, 453592.37);
      expect(testMass.inGrams, 453.59237);
      expect(testMass.inKilograms, 0.45359237);
      expect(testMass.inMetricTonnes, 0.00045359237);
      expect(testMass.inOunces, 16);
      expect(testMass.inStones, 0.07142857142857142);
      expect(testMass.inPounds, 1);
      expect(testMass.inImperialTonnes, 0.0004464285714285714);
    });
  });
}


test/length_test.dart
import 'package:measurements/measurements.dart';
import 'package:test/test.dart';

void main() {
  Length testLength;

  group("Length", () {
    test("fromNanometres", () {
      testLength = Length.fromNanometres(304800000);

      expect(testLength.inMicrometres, 304800);
      expect(testLength.inMillimetres, 304.8);
      expect(testLength.inCentimetres, 30.48);
      expect(testLength.inMetres, 0.3048);
      expect(testLength.inKilometres, 0.0003048);
      expect(testLength.inInches, 12);
      expect(testLength.inFeet, 1);
      expect(testLength.inYards, 1 / 3);
      expect(testLength.inMiles, 0.0001893939393939394);
      expect(testLength.inNauticalMiles, 0.00016457883369330455);
    });

    test("fromMicrometres", () {
      testLength = Length.fromMicrometres(304800);

      expect(testLength.inNanometres, 304800000);
      expect(testLength.inMillimetres, 304.8);
      expect(testLength.inCentimetres, 30.48);
      expect(testLength.inMetres, 0.3048);
      expect(testLength.inKilometres, 0.0003048);
      expect(testLength.inInches, 12);
      expect(testLength.inFeet, 1);
      expect(testLength.inYards, 1 / 3);
      expect(testLength.inMiles, 0.0001893939393939394);
      expect(testLength.inNauticalMiles, 0.00016457883369330455);
    });

    test("fromMillimetres", () {
      testLength = Length.fromMillimetres(304.8);

      expect(testLength.inNanometres, 304800000);
      expect(testLength.inMicrometres, 304800);
      expect(testLength.inCentimetres, 30.48);
      expect(testLength.inMetres, 0.3048);
      expect(testLength.inKilometres, 0.0003048);
      expect(testLength.inInches, 12);
      expect(testLength.inFeet, 1);
      expect(testLength.inYards, 1 / 3);
      expect(testLength.inMiles, 0.0001893939393939394);
      expect(testLength.inNauticalMiles, 0.00016457883369330455);
    });

    test("fromCentimetres", () {
      testLength = Length.fromCentimetres(30.48);

      expect(testLength.inNanometres, 304800000);
      expect(testLength.inMicrometres, 304800);
      expect(testLength.inMillimetres, 304.8);
      expect(testLength.inMetres, 0.3048);
      expect(testLength.inKilometres, 0.0003048);
      expect(testLength.inInches, 12);
      expect(testLength.inFeet, 1);
      expect(testLength.inYards, 1 / 3);
      expect(testLength.inMiles, 0.0001893939393939394);
      expect(testLength.inNauticalMiles, 0.00016457883369330455);
    });

    test("fromMetres", () {
      testLength = Length.fromMetres(0.3048);

      expect(testLength.inNanometres, 304800000);
      expect(testLength.inMicrometres, 304800);
      expect(testLength.inMillimetres, 304.8);
      expect(testLength.inCentimetres, 30.48);
      expect(testLength.inKilometres, 0.0003048);
      expect(testLength.inInches, 12);
      expect(testLength.inFeet, 1);
      expect(testLength.inYards, 1 / 3);
      expect(testLength.inMiles, 0.0001893939393939394);
      expect(testLength.inNauticalMiles, 0.00016457883369330455);
    });

    test("fromKilometres", () {
      testLength = Length.fromKilometres(0.0003048);

      expect(testLength.inNanometres, 304800000);
      expect(testLength.inMicrometres, 304800);
      expect(testLength.inMillimetres, 304.8);
      expect(testLength.inCentimetres, 30.48);
      expect(testLength.inMetres, 0.3048);
      expect(testLength.inInches, 12);
      expect(testLength.inFeet, 1);
      expect(testLength.inYards, 1 / 3);
      expect(testLength.inMiles, 0.0001893939393939394);
      expect(testLength.inNauticalMiles, 0.00016457883369330455);
    });

    test("fromInches", () {
      testLength = Length.fromInches(12);

      expect(testLength.inNanometres, 304800000);
      expect(testLength.inMicrometres, 304800);
      expect(testLength.inMillimetres, 304.8);
      expect(testLength.inCentimetres, 30.48);
      expect(testLength.inMetres, 0.3048);
      expect(testLength.inKilometres, 0.0003048);
      expect(testLength.inFeet, 1);
      expect(testLength.inYards, 1 / 3);
      expect(testLength.inMiles, 0.0001893939393939394);
      expect(testLength.inNauticalMiles, 0.00016457883369330455);
    });

    test("fromFeet", () {
      testLength = Length.fromFeet(1);

      expect(testLength.inNanometres, 304800000);
      expect(testLength.inMicrometres, 304800);
      expect(testLength.inMillimetres, 304.8);
      expect(testLength.inCentimetres, 30.48);
      expect(testLength.inMetres, 0.3048);
      expect(testLength.inKilometres, 0.0003048);
      expect(testLength.inInches, 12);
      expect(testLength.inYards, 1 / 3);
      expect(testLength.inMiles, 0.0001893939393939394);
      expect(testLength.inNauticalMiles, 0.00016457883369330455);
    });

    test("fromYards", () {
      testLength = Length.fromYards(1 / 3);

      expect(testLength.inNanometres, 304800000);
      expect(testLength.inMicrometres, 304800);
      expect(testLength.inMillimetres, 304.8);
      expect(testLength.inCentimetres, 30.48);
      expect(testLength.inMetres, 0.3048);
      expect(testLength.inKilometres, 0.0003048);
      expect(testLength.inInches, 12);
      expect(testLength.inFeet, 1);
      expect(testLength.inMiles, 0.0001893939393939394);
      expect(testLength.inNauticalMiles, 0.00016457883369330455);
    });

    test("fromMiles", () {
      testLength = Length.fromMiles(0.0001893939393939394);

      expect(testLength.inNanometres, 304800000);
      expect(testLength.inMicrometres, 304800);
      expect(testLength.inMillimetres, 304.8);
      expect(testLength.inCentimetres, 30.48);
      expect(testLength.inMetres, 0.3048);
      expect(testLength.inKilometres, 0.0003048);
      expect(testLength.inInches, 12);
      expect(testLength.inFeet, 1);
      expect(testLength.inYards, 1 / 3);
      expect(testLength.inNauticalMiles, 0.00016457883369330455);
    });

    test("fromNauticalMiles", () {
      testLength = Length.fromNauticalMiles(0.00016457883369330455);

      expect(testLength.inNanometres, 304800000);
      expect(testLength.inMicrometres, 304800);
      expect(testLength.inMillimetres, 304.8);
      expect(testLength.inCentimetres, 30.48);
      expect(testLength.inMetres, 0.3048);
      expect(testLength.inKilometres, 0.0003048);
      expect(testLength.inInches, 12);
      expect(testLength.inFeet, 1);
      expect(testLength.inYards, 1 / 3);
      expect(testLength.inMiles, 0.0001893939393939394);
    });
  });
}


lib/src/area.dart
part of measurements;

/// An area [Measurement] - for example 55.67m².
///
/// To create a new `Area` object, use any of this class's factory
/// constructors.
/// For example,
/// ```dart
/// Area plotArea = Area.fromHectares(20.0);
/// ```
///
/// Use one of the properties, such as [inSquareFeet], to retrieve the double value
/// of the `Area` in the specified Area unit.
/// For example,
/// ```dart
/// Area plotArea = Area.fromHectares(20.0);
/// assert(plotArea.fromSquareMetres == 200000.0);
/// ```
/// This class provides a collection of arithmetic and comparison operators.
/// See [Measurement] for the collection of arithmetic and comparison operators
/// which this class provides.
class Area extends Measurement<Area> {
  /// A zero-valued 'Area' constant.
  static const Area zero = const Area.fromSquareCentimetres(0);

  static const double _squareCentimetrePerSquareMetre = 10000;
  static const double _squareCentimetrePerSquareKilometre = 10000000000;
  static const double _squareCentimetrePerSquareInch = 16129 / 2500;
  static const double _squareCentimetrePerSquareFoot = 580644 / 625;
  static const double _squareCentimetrePerSquareYard = 5225796 / 625;
  static const double _squareCentimetrePerSquareMile = 647497027584 / 25;
  static const double _squareCentimetrePerHectare = 100000000;
  static const double _squareCentimetrePerAcre = 40468564.224;

  // Returns a new `Area` object using baseValue as the Area in cm².
  @override
  Area _getInstance(double baseValue) {
    return Area.fromSquareCentimetres(baseValue);
  }

  /// Returns a new `Area` object from the length and width of a rectangle.
  factory Area.fromRectangle(Length length, Length width) {
    return Area.fromSquareCentimetres(
        length.inCentimetres * width.inCentimetres);
  }

  /// Returns a new `Area` object from the base and height of a triangle.
  factory Area.fromTriangle(Length base, Length height) {
    return Area.fromSquareCentimetres(
        base.inCentimetres * height.inCentimetres / 2);
  }

  /// Returns a new `Area` object from the radius of a circle.
  factory Area.fromCircle(Length radius) {
    return Area.fromSquareCentimetres(
        math.pi * math.pow(radius.inCentimetres, 2));
  }

  /// Returns a new `Area` object from the number of cm².
  const Area.fromSquareCentimetres(double squareCentimetres)
      : super(squareCentimetres);

  /// Returns a new `Area` object from the number of Centimetres.
  const Area.fromSquareMetres(double squareMetres)
      : super(squareMetres * _squareCentimetrePerSquareMetre);

  /// Returns a new `Area` object from the number of centimetres.
  const Area.fromSquareKilometres(double squareKilometres)
      : super(squareKilometres * _squareCentimetrePerSquareKilometre);

  /// Returns a new `Area` object from the number of metres.
  const Area.fromSquareFeet(double squareFeet)
      : super(squareFeet * _squareCentimetrePerSquareFoot);

  /// Returns a new `Area` object from the number of kilometres.
  const Area.fromSquareInches(double squareInches)
      : super(squareInches * _squareCentimetrePerSquareInch);

  /// Returns a new `Area` object from the number of inches.
  const Area.fromHectares(double hectares)
      : super(hectares * _squareCentimetrePerHectare);

  /// Returns a new `Area` object from the number of feet.
  const Area.fromAcres(double acres) : super(acres * _squareCentimetrePerAcre);

  /// Returns a new `Area` object from the number of yards.
  const Area.fromSquareYards(double squareYards)
      : super(squareYards * _squareCentimetrePerSquareYard);

  /// Returns a new `Area` object from the number of miles.
  const Area.fromSquareMiles(double squareMiles)
      : super(squareMiles * _squareCentimetrePerSquareMile);

  /// Returns the total number of centimetres making up this `Area`.
  double get inSquareCentimetres => _baseValue;

  /// Returns the total number of metres² making up this `Area`.
  double get inSquareMetres => _baseValue / _squareCentimetrePerSquareMetre;

  /// Returns the total number of kilometres² making up this `Area`.
  double get inSquareKilometres =>
      _baseValue / _squareCentimetrePerSquareKilometre;

  /// Returns the total number of feet² making up this `Area`.
  double get inSquareFeet => _baseValue / _squareCentimetrePerSquareFoot;

  /// Returns the total number of km² making up this `Area`.
  double get inSquareInches => _baseValue / _squareCentimetrePerSquareInch;

  /// Returns the total number of hectares making up this `Area`.
  double get inHectares => _baseValue / _squareCentimetrePerHectare;

  /// Returns the total number of acres making up this `Area`.
  double get inAcres => _baseValue / _squareCentimetrePerAcre;

  /// Returns the total number of yards² making up this `Area`.
  double get inSquareYards => _baseValue / _squareCentimetrePerSquareYard;

  /// Returns the total number of miles² making up this `Area`.
  double get inSquareMiles => _baseValue / _squareCentimetrePerSquareMile;
}


lib/src/measurement.dart
part of measurements;

/// Base class used by types that represent forms of measurement.
///
/// This class provides all `Measurement` subclasses with a collection of
/// arithmetic and comparison operators.
///
/// See also:
/// [Length], for measurements of length.
/// [Mass], for measurements of weight.
/// [Area], for measurements of area.
abstract class Measurement<T extends Measurement<T>> implements Comparable<T> {
  // Represents the value of this measurement in the smallest possible units.
  // It is used for the implementation of all arithmetic and comparison operators.
  final double _baseValue;

  // Called by subclasses to instantiate _baseValue
  const Measurement(this._baseValue);

  // The abstract factory method that must be implemented by all subclasses.
  T _getInstance(double baseValue);

  /// Adds this `Measurement` and [other] and
  /// returns the sum as a new object of [T].
  T operator +(T other) => _getInstance(_baseValue + other._baseValue);

  /// Subtracts [other] from this `Measurement` and
  /// returns the difference as a new object of [T].
  T operator -(T other) => _getInstance(_baseValue - other._baseValue);

  /// Multiples this `Measurement` by [other] and
  /// returns the result as a new object of [T].
  T operator *(double other) => _getInstance(_baseValue * other);

  /// Divides this `Measurement` by [other] and
  /// returns the result as a new object of [T].
  T operator /(double other) => _getInstance(_baseValue / other);

  /// Returns a new object of [T] representing this `Measurement` negated.
  ///
  /// The returned `Measurement` has the same length as this one, but will have the
  /// opposite sign of this one.
  T operator -() => _getInstance(0 - _baseValue);

  /// Returns `true` if the value of this `Measurement`
  /// is less than the value of [other].
  bool operator <(T other) => _baseValue < other._baseValue;

  /// Returns `true` if the value of this `Measurement`
  /// is greater than the value of [other].
  bool operator >(T other) => _baseValue > other._baseValue;

  /// Returns `true` if the value of this `Measurement`
  /// is less than or equal to the value of [other].
  bool operator <=(T other) => _baseValue <= other._baseValue;

  /// Returns `true` if the value of this `Measurement`
  /// is greater than or equal to the value of [other].
  bool operator >=(T other) => _baseValue >= other._baseValue;

  /// Returns `true` if this [Measurement] has the same value as [other].
  @override
  bool operator ==(dynamic other) =>
      other is T && (_baseValue == other._baseValue);

  @override
  int get hashCode => _baseValue.hashCode;

  /// Compares this `Measurement` to [other], returning zero if the values are
  /// equal.
  ///
  /// Returns a negative integer if this `Measurement` is less than
  /// [other], or a positive integer if it is greater.
  ///
  /// A negative `Measurement` is always considered shorter than a positive one.
  ///
  /// It is always the case that `measurement1.compareTo(measurement2) < 0` iff
  /// `(someMeasurement + measurement1).compareTo(someMeasurement + measurement2) < 0`.
  @override
  int compareTo(T other) => _baseValue.compareTo(other._baseValue);

  /// Returns a new object of [T] representing the absolute value of this
  /// `Measurement`.
  ///
  /// The returned object has the same base value as this one, but is always
  /// positive.
  T abs() => _getInstance(_baseValue.abs());

  /// Returns whether this `Measurement` is negative.
  bool get isNegative => _baseValue < 0;
}


example/main.dart
import 'package:measurements/measurements.dart';

void main(List<String> arguments) async {
  print("Some interesting measurements...");

  // Conversions
  Length heightOfEverest = Length.fromMetres(8848.0);
  print("Height of Everest (in metres): ${heightOfEverest.inMetres}.");
  print("Height of Everest (in kilometres): ${heightOfEverest.inKilometres}.");
  print("Height of Everest (in yards): ${heightOfEverest.inYards}.");
  print("Height of Everest (in miles): ${heightOfEverest.inMiles}.");

  // Calculations
  Area a = Area.fromAcres(87.0);
  Area b = Area.fromAcres(43.0);
  print("a + b is ${(a + b).inAcres} acres.");
  print("a - b is ${(a - b).inAcres} acres.");
  print("a * 4 is ${(a * 4).inAcres} acres.");
  print("a / 4 is ${(a / 4).inAcres} acres.");

  // Comparisons
  Mass elephantWeight = Mass.fromKilograms(5400);
  Mass zebraWeight = Mass.fromKilograms(380);
  if (elephantWeight == zebraWeight) {
    print("An elephant and a Zebra have the same weight.");
  } else if (zebraWeight < elephantWeight) {
    Mass difference = elephantWeight - zebraWeight;
    print("An elephant is ${difference.inPounds} pounds heavier than a zebra.");
  } else {
    Mass difference = zebraWeight - elephantWeight;
    print("A Zebra is ${difference.inPounds} pounds heavier than an elephant.");
  }
}
